Free ByteBuffers in tests and fix some bugs.

gianm opened the pull request:
Many tests had problems where a direct ByteBuffer would be allocated
and then not freed. This is bad because it causes flaky tests.

To fix this:

1) Add ByteBufferUtils.allocateDirect(size), which returns a ResourceHolder.
   This makes it easy to free the direct buffer. Currently, it's only used
   in tests, because production code seems OK.

2) Update all usages of ByteBuffer.allocateDirect (off-heap) in tests either
   to ByteBuffer.allocate (on-heap, which are garbaged collected), or to
   ByteBufferUtils.allocateDirect (wherever it seemed like there was a good
   reason for the buffer to be off-heap). Make sure to close all direct
   holders when done.

This patch fixes three other issues that came up while debugging flaky tests:

1) Various places in the tests altered the global NullHandlingConfig but did not
    reset it. Now, they do properly reset it.

2) ExpressionLambdaAggregatorFactory's getComparator, combine, and
    finalizeComputation methods were not thread-safe. AggregatorFactory
    instances are shared across threads at various points, so they do need
    to be thread-safe. Now, they're made thread-safe by moving bindings
    from Suppliers to ThreadLocals.

3) ExpressionLambdaBufferAggregator's get method would sometimes retain a
    reference to the underlying buffer, which isn't allowed per contract of
    BufferAggregator. Now, it's copied when necessary.

kfaraz commented on pom.xml:
```diff
@@ -1733,7 +1733,7 @@
                             <argLine>
                                 ${jdk.surefire.argLine}
-                                -Xmx768m -Duser.language=en -Duser.country=US -Dfile.encoding=UTF-8
+                                -Xmx1280m -Duser.language=en -Duser.country=US -Dfile.encoding=UTF-8
```
Is this needed because several tests are now going to use on-heap byte buffers?
Do those tests fail without this change?

gianm replied:
I don't know if this is needed. Right now I'm wrestling with inexplicable heap-usage errors early in the job cycle, like the following one. I thought this change might help but it doesn't. I need to do some more investigation into where this heap usage is coming from.

```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.druid.guice.MetadataStorageTablesConfigTest
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 7.333 s <<< FAILURE! - in org.apache.druid.guice.MetadataStorageTablesConfigTest
[ERROR] org.apache.druid.guice.MetadataStorageTablesConfigTest.testSerdeMetadataStorageTablesConfig  Time elapsed: 7.076 s  <<< ERROR!
java.lang.OutOfMemoryError: GC overhead limit exceeded
	at java.util.zip.ZipCoder.getBytes(ZipCoder.java:80)
  	at java.util.zip.ZipFile.getEntry(ZipFile.java:316)
  	at java.util.jar.JarFile.getEntry(JarFile.java:240)
  	at java.util.jar.JarFile.getJarEntry(JarFile.java:223)
  	at sun.misc.URLClassPath$JarLoader.getResource(URLClassPath.java:1054)
  	at sun.misc.URLClassPath.getResource(URLClassPath.java:249)
  	at java.net.URLClassLoader$1.run(URLClassLoader.java:366)
  	at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
  	at java.security.AccessController.doPrivileged(Native Method)
  	at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
  	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
  	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
  	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
  	at java.lang.ClassLoader.defineClass1(Native Method)
  	at java.lang.ClassLoader.defineClass(ClassLoader.java:756)
  	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
  	at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
  	at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
  	at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
  	at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
  	at java.security.AccessController.doPrivileged(Native Method)
  	at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
  	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
  	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
  	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
  	at org.hibernate.validator.HibernateValidator.createGenericConfiguration(HibernateValidator.java:31)
  	at javax.validation.Validation$GenericBootstrapImpl.configure(Validation.java:276)
  	at javax.validation.Validation.buildDefaultValidatorFactory(Validation.java:110)
  	at org.apache.druid.guice.ConfigModule.configure(ConfigModule.java:39)
  	at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
  	at org.apache.druid.guice.MetadataStorageTablesConfigTest$1.configure(MetadataStorageTablesConfigTest.java:48)
  	at com.google.inject.spi.Elements$RecordingBinder.install(Elements.java:340)
```

gianm replied:
I was able to reproduce OOMEs on processing tests locally. I think that the processing and merge buffers for groupBy tests need to stay off-heap, because there's a bunch of them and they're 10MB each. It adds up. I'll try organizing things so that they stay offheap but get freed properly.

kfaraz commented on core/src/main/java/org/apache/druid/java/util/common/ByteBufferUtils.java:
```diff
@@ -138,6 +144,45 @@ private static MethodHandle unmapJava9(MethodHandles.Lookup lookup) throws Refle
     return unmapper.bindTo(UnsafeUtils.theUnsafe());
   }
 
+  /**
```
Could we also include a comment advising when `allocateDirect` should be used (especially in tests) rather than just using on-heap buffers with `ByteBuffer.allocate()`?

gianm replied:
Sure, I'll add the following. Let me know what you think.

```
  /**
   * Same as {@link ByteBuffer#allocateDirect(int)}, but returns a closeable {@link ResourceHolder} that
   * frees the buffer upon close.
   *
   * Direct (off-heap) buffers are an alternative to on-heap buffers that allow memory to be managed
   * outside the purview of the garbage collector. It's most useful when allocating big chunks of memory,
   * like processing buffers.
   */
```

kfaraz replied:
It looks good, thanks!

kfaraz commented on core/src/main/java/org/apache/druid/java/util/common/ByteBufferUtils.java:
```diff
@@ -138,6 +144,45 @@ private static MethodHandle unmapJava9(MethodHandles.Lookup lookup) throws Refle
+   * Same as {@link ByteBuffer#allocateDirect(int)}, but returns a closeable {@link ResourceHolder} that
+   * frees the buffer upon close.
+   */
+  public static ResourceHolder<ByteBuffer> allocateDirect(final int size)
```
Should we also use this utility method in places such as `LZ4Compressor.allocateInBuffer()` rather than calling `free` directly from there?

kfaraz reviewed (commented):
Thanks, @gianm, tests are much cleaner now!

gianm commented:
Updated NestedQueryPushDownTest as well: it now uses off-heap buffers (because there's 20x 10MB buffers: quite big) and properly frees them. Also, added `-XX:+ExitOnOutOfMemoryError -XX:+HeapDumpOnOutOfMemoryError` to all surefire argLines, to aid in debugging any OOMEs that may occur in the future.

clintropolis reviewed (approved):
+1 after ci

gianm commented:
My current guess about the `ScanQueryRunnerTest.testFullOnSelectAsCompactedList` problem is that null-handling config isn't always getting reset properly in tests. I found a few places where it wasn't being reset, and updated them all in the latest commit.

I'm still not sure what's going on with `GroupByQueryRunnerTest.testGroupByWithExpressionAggregatorWithComplex`.

gianm commented:
My current guess about `GroupByQueryRunnerTest.testGroupByWithExpressionAggregatorWithComplex` is that the problem is ExpressionLambdaAggregatorFactory's getComparator, combine, and finalizeComputation methods are not thread-safe. I replaced the Suppliers with ThreadLocals.

gianm commented:
The commit https://github.com/apache/druid/pull/12521/commits/8b17d2708b6a6f8804d182bd5a7431037d0ef2db fixes an issue where ExpressionLambdaBufferAggregator.get would retain references to the underlying ByteBuffer. This could believably cause the crashing issues I've been seeing. 🤞 for CI…

clintropolis reviewed (approved):
good finds, and thanks for cleaning up my mess :metal:
